ramiro_serra/Debug and profiling notes.md

```diff
@@ -0,0 +1,113 @@
+    
+    <launch>
+        <arg name="x" default="2.0"/>
+        <arg name="y" default="2.0"/>
+        <arg name="Y" default="1.0"/>
+        <arg name="ns" default="create1"/>
+        <node pkg="ca_move_base" type="send_goal" name="navigation_goals"
+                args="$(arg x) $(arg y) $(arg Y)" ns="$(arg ns)" launch-prefix="valgrind --tool=callgrind --callgrind-out-file='callgrind.sendgoal.%p'"/>
+    </launch>
+
+Note i added some default values to the required arguments, only to make the usage of the launchfile more straight-forward.
+We can launch now the file with `roslaunch ca_move_base send_goal.launch`, we can let it run a little and then `C-c` to stop it.
+Valgrind will automatically create a log located in `~/.ros/`, let's check it:
+    
+    cd ~/.ros/
+    ls
+
+Then we can open it with `kcachegrind`, that we installed earlier
+    
+    kcachegrind callgrind.sendgoal.10303
+*The appended number may very with the PID associated.*
+
+![kcachegrind](media/kcachegrind.png)
+
+Now a window shows up, with a lot of info of the program, if you want to profundize more on the topic, you can see [here](http://valgrind.org/docs/), and learn how to interpret the data.
+
+## Debugging with gdb
+
+The process is pretty similar to the previous case, but in this one, the `launch-prefix` differs a little, let's see:
+
+We have some options depending on how we want to launch gdb, in my case, I chose the *run your node in a gdb in a separate xterm window, manually type run to start it* way.
+So, we have to change the launch prefix of `send_goal` to match with the info on the wiki page.
+This is how my `send_goal.launch` file looks like:
+
+    <launch>
+        <arg name="x" default="2.0"/>
+        <arg name="y" default="2.0"/>
+        <arg name="Y" default="1.0"/>
+
+        <arg name="ns" default="create1"/>
+
+        <node pkg="ca_move_base" type="send_goal" name="navigation_goals"
+                args="$(arg x) $(arg y) $(arg Y)" ns="$(arg ns)" launch-prefix="xterm -e gdb --args"/>
+    </launch>
+
+We can launch now the file with `roslaunch ca_move_base send_goal.launch`, and a `xterm` window will popup.
+We can now use every functionality that `gdb` offers, I recommend seeing [this cheatsheet](https://darkdust.net/files/GDB%20Cheat%20Sheet.pdf) for some useful commands.
+
+Let's see a quick example, first set a setpoint at line 10 of the file, `b 10`.
+The output:
+
+    Breakpoint 1 at 0x47ee62: file /create_ws/src/create_autonomy/navigation/ca_move_base/src/send_robot_goal.cpp, line 10.
+    (gdb) 
+
+Then run the program with `run`.
+We can see the program stopped at the checkpoint we set previously.
+
+    Breakpoint 1, __static_initialization_and_destruction_0 (__initialize_p=1, 
+        __priority=65535)
+        at /create_ws/src/create_autonomy/navigation/ca_move_base/src/send_robot_goal.cpp:10
+    10      const std::string NODE_NAME = "navigation_goals";
+
+We could watch a variable with `watch`, and many more features.

```

Comment:
what do you see if you run `where full`?

Reply:
Added